### PR TITLE
Python: Add agent function termination sample. Fix openai chat completion chat history bug.

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

The Python agent samples were missing a concept sample to show how to use the auto function invocation filter with a ChatCompletionAgent. It's similar to how the filter is used with a normal chat completion service; however, it's still good to show in the context of using agents. 

Additionally, during sample creation, it was found that the chat history message, during non-streaming function calling, was always being added, even if the model did not return a function to call. This was causing a duplication of chat history messages as the caller is usually responsible to add the message to the chat history once it receives it from the AI connector.

### Description

This PR:
- Adds the auto function invoke chat completion agent sample
- Fixes the chat history bug by re-ordering when the tool call message from the model is added to the chat history during a non-streaming `get_chat_message_contents` call-- it is added after we find that we do have functions to call. Otherwise, do not add the completions example to the chat history as the caller is responsible for doing that.
   - Fixed for AzureAI Inference Chat Completion, Google VertexAI Chat Completion, and GoogleAI Chat Completion
- Fixes the auto invoke function chat message content construction bug for other AI connectors that was fixed for OpenAI in #8098

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

**Comment by markwallace-microsoft** [comment]
## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 2370 | 1 :zzz: | 0 :x: | 0 :fire: | 56.650s :stopwatch: |

File: python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py [hunk 1]
```diff
@@ -109,13 +109,15 @@ async def get_chat_message_contents(
...
-            function_calls = [item for item in chat_history.messages[-1].items if isinstance(item, FunctionCallContent)]
+            # get the function call contents from the chat message, there is only one chat message
+            # this was checked earlier
+            function_calls = [item for item in completions[0].items if isinstance(item, FunctionCallContent)]
             if (fc_count := len(function_calls)) == 0:
                 return completions
 
+            # Since we have a function call, add the assistant's tool call message to the history
+            chat_history.add_message(message=completions[0])
+
```
Comment (TaoChenOSU) [inline python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:120]: If we decided that this is the right fix, we need to fix it for all connectors.

Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:120]: I'd prefer to keep this scoped to OpenAI, and then we can create a PR for the other connectors as needed.

Comment (TaoChenOSU) [reply python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:120]: That sounds good to me! Should we also introduce another PR that fixes the issue in the `else` clause where function calling is not turned off?

**Review by TaoChenOSU — APPROVED** [review]

File: python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py (hunk 1 above)
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:121]: I decided to fix it for all connectors affected in this PR. Please have a look.

**Review by crickman — APPROVED** [review]

**Review by eavanvalkenburg — APPROVED** [review]

File: python/tests/integration/completions/test_chat_completion_with_function_calling.py [hunk 2]
```diff
@@ -543,7 +543,7 @@ async def _test_helper(
             retries=5,
         )
 
-        if test_type != FunctionChoiceTestTypes.AUTO or stream:
+        if test_type != stream:
```
Comment (TaoChenOSU) [inline python/tests/integration/completions/test_chat_completion_with_function_calling.py:546]: Is this intended? `stream` is a bool while test_type is an enum.

**Review by eavanvalkenburg — APPROVED** [review]

File: python/tests/integration/completions/test_chat_completion_with_function_calling.py (hunk 2 above)
Comment (moonbox3) [reply python/tests/integration/completions/test_chat_completion_with_function_calling.py]: Good catch fixed!

**Review by TaoChenOSU — APPROVED** [review]

**Pull request merged**